### ISSUE TITLE
number format pattern

### ISSUE THREAD

**Issue by hongsudt** [informatics-isi-edu/ermrestjs#299]
@karl: I need more than two digits of fraction in the floating point display. My project has small values which are being rounded and truncated too much and giving poor UX.

In the long run, the pre_format display option is meant to allow this to be customized. I know there are some concerns about which formatting language and library to use here.

@hongsudt commented on [Wed Nov 02 2016](https://github.com/informatics-isi-edu/ermrest/issues/115)

We need to start working on the template. I hear quite a few complaints about the GUI presenting an ID column in a number format (e.g. with , separating thousands). 

@karlcz @robes is there any js number formatting template that we can use? 

@hongsudt not sure what you mean by js number formatting template.

Sounds like if the concern is not formatting the number when it is a ID column, maybe we either need a heuristic for that, or we need a turn off formatting annotation?

@karlcz commented on [Fri Dec 02 2016](https://github.com/informatics-isi-edu/ermrest/issues/115#issuecomment-264608146)

I think this issue is mostly an ermrestjs issue?  We need to pick a posix-style formatting library so we can implement the pre_format display annotation. Then we can update the annotations doc with more precise dialect info for the format directives.

**Comment by chiragsanghvi** [comment]
@hongsudt @karlcz Will this format work

The placeholders in the format string are marked by `%` and are followed by one or more of these elements, in this order:
- An optional `+` sign that forces to preceed the result with a plus or minus sign on numeric values. By default, only the - sign is used on negative numbers.
- An optional padding specifier that says what character to use for padding (if specified). Possible values are `0` or any other character precedeed by a `'` (single quote). The default is to pad with spaces.
- An optional `-` sign, that causes `printf` to left-align the result of this placeholder. The default is to right-align the result.
- An optional number, that says how many characters the result should have. If the value to be returned is shorter than this number, the result will be padded. When used with the `j` (JSON) type specifier, the padding length specifies the tab size used for indentation.
- An optional precision modifier, consisting of a `.` (dot) followed by a number, that says how many digits should be displayed for floating point numbers. When used with the `g` type specifier, it specifies the number of significant digits. When used on a string, it causes the result to be truncated.
- A type specifier that can be any of:
```
% — yields a literal % character
b — yields an integer as a binary number
c — yields an integer as the character with that ASCII value
d or i — yields an integer as a signed decimal number
e — yields a float using scientific notation
u — yields an integer as an unsigned decimal number
f — yields a float as is; see notes on precision above
g — yields a float as is; see notes on precision above
o — yields an integer as an octal number
s — yields a string as is
t — yields true or false
X — yields an integer as a hexadecimal number (upper-case)
j — yields a JavaScript object or array as a JSON encoded string
```

Examples
```
printf('%%') => %
printf('%b', 2) => 10
printf('%c', 65) => A
printf('%d', 2) => 2
printf('%i', 2) => 2
printf('%d', '2') => 2
printf('%i', '2') => 2
printf('%j', [object Object]) => {"foo":"bar"}
printf('%j', foo,bar) => ["foo","bar"]
printf('%e', 2) => 2e+0
printf('%u', 2) => 2
printf('%u', -2) => 4294967294
printf('%f', 2.2) => 2.2
printf('%g', 3.141592653589793) => 3.141592653589793
printf('%o', 8) => 10
printf('%o', -8) => 37777777770
printf('%s', '%s') => %s
printf('%X', 255) => FF
printf('%X', -255) => FFFFFF01
printf('%2$s %3$s a %1$s', 'cracker') => Polly wants a cracker
printf('Hello %(who)s!', [object Object]) => Hello world!
printf('%t', true) => true
printf('%.1t', true) => t
printf('%t', 'true') => true
printf('%t', 1) => true
printf('%t', false) => false
printf('%.1t', false) => f
printf('%t', '') => false
printf('%t', 0) => false
printf('%d', 2) => 2
printf('%d', -2) => -2
printf('%+d', 2) => +2
printf('%+d', -2) => -2
printf('%i', 2) => 2
printf('%i', -2) => -2
printf('%+i', 2) => +2
printf('%+i', -2) => -2
printf('%f', 2.2) => 2.2
printf('%f', -2.2) => -2.2
printf('%+f', 2.2) => +2.2
printf('%+f', -2.2) => -2.2
printf('%+.1f', -2.34) => -2.3
printf('%+.1f', -0.01) => -0.0
printf('%.6g', 3.141592653589793) => 3.14159
printf('%.3g', 3.141592653589793) => 3.14
printf('%.1g', 3.141592653589793) => 3
printf('%+010d', -123) => -000000123
printf('%+'_10d', -123) => ______-123
printf('%f %f', -234.34) => -234.34 123.2
printf('%05d', -2) => -0002
printf('%05i', -2) => -0002
printf('%5s', '<') =>     <
printf('%05s', '<') => 0000<
printf('%'_5s', '<') => ____<
printf('%-5s', '>') => >    
printf('%0-5s', '>') => >0000
printf('%'_-5s', '>') => >____
printf('%5s', 'xxxxxx') => xxxxxx
printf('%02u', 1234) => 1234
printf('%8.3f', -10.23456) =>  -10.235
printf('%f %s', -12.34) => -12.34 xxx
printf('%2j', {foo: 'bar'}) => {\n  "foo": "bar"\n}
printf('%2j', foo,bar) => [\n  "foo",\n  "bar"\n]
printf('%.1f', 2.345) => 2.3
printf('%5.5s', 'xxxxxx') => xxxxx
printf('%5.1s', 'xxxxxx') =>     x
```

**Comment by RFSH** [comment]
Isn't this syntax supposed to be something that we want to add to the current pattern language that we have in mustache? Or are we completely moving away from mustache and using this syntax?

**Comment by chiragsanghvi** [comment]
This syntax will be used on top of Mustache and is valid for one column value only. You won't be able to access other column values while  formatting. 

if a column has pre_format annotation then instead of using the default format function of ermrestjs we format it according to above syntax allowing user to provide precision and other stuff.

Once formatted this value will be provided to the Mustache template of all columns.

**Comment by hongsudt** [comment]
@chiragsanghvi What library are we using for this? 

Also, there doesn't seem to be an example on how to enable thousand separating tokens (e.g. `,` in most countries or `.` in some countries), so we will see `1,000,000` instead of `1000000`.

Also, are you suggesting that we only have to specify the first part of the printf syntax?

**Comment by karlcz** [comment]
This seems good to me.  Is there a printf in javascript now or is this a third-party library?

When I saw the part accessing JSON fields I thought, man that seems better than the crazy mustache syntax. :-)  But then I am sure we still have to struggle with the conditionals, looping, and null guards...

**Comment by chiragsanghvi** [comment]
We are using a library called [sprintfjs](https://www.npmjs.com/package/sprintf-js). 

All other formatting can be added as I have made changes to the library according to my requirement

**Comment by karlcz** [comment]
Oh, does the padding option for JSON turn on pretty printing, i.e. multi-line format?

**Comment by chiragsanghvi** [comment]
@karlcz This is a small 200 line third party library, but to customize to our needs I have made some changes to it.

**Comment by chiragsanghvi** [comment]
@karlcz We might be able add pretty printing by introducing one more option.

**Comment by hongsudt** [comment]
@chiragsanghvi I am not sure you have answered my question about the thousand separator token.. If we don't already have this, we should implement it.. We will continue to use our existing default if there is no annotation.

**Comment by hongsudt** [comment]
It seems like the thousand separator is accomplished by setting locale in C. Should we follow the same approach? 

```
   setlocale(LC_NUMERIC, "");
    printf("%'d\n", 1123456789);
```

**Comment by RFSH** [comment]
@chiragsanghvi has implemented this feature in #564. He also wrote [this](https://github.com/informatics-isi-edu/ermrestjs/wiki/Pre-Format-Annotation) wiki page for it.
